src/redisearch_rs/slots_tracker/src/slots_tracker.rs

```diff
@@ -0,0 +1,701 @@
+    local: SlotSet,
+    /// Fully available non-owned slots - locally available but not owned by this instance.
+    fully_available: SlotSet,
+    /// Partially available non-owned slots - partially available and not owned by this instance.
+    partially_available: SlotSet,
+    /// Version counter for tracking changes to the slots configuration.
+    /// Incremented whenever the slot configuration changes. Wraps around safely.
+    version: u32,
+}
+
+impl SlotsTracker {
+    /// Creates a new SlotsTracker with empty slot sets and version 0.
+    pub const fn new() -> Self {
+        Self {
+            local: SlotSet::new(),
+            fully_available: SlotSet::new(),
+            partially_available: SlotSet::new(),
+            version: 0,
+        }
+    }
+
+    /// Gets the current version counter value.
+    pub const fn get_version(&self) -> u32 {
+        self.version
+    }
+
+    /// Increments the version counter, skipping reserved values.
+    ///
+    /// If the current version is `MAX_VALID_VERSION` (u32::MAX - 2), wraps to 0
+    /// instead of incrementing to reserved values (u32::MAX - 1 or u32::MAX).
+    fn increment_version(&mut self) {
+        debug_assert!(self.version <= MAX_VALID_VERSION);
+        self.version = if self.version < MAX_VALID_VERSION {
+            self.version + 1
+        } else {
+            0 // Wrap around to 0
+        };
+    }
+
+    /// Sets the local responsibility slot ranges.
+    ///
+    /// This function replaces the `local` set with the provided ranges.
+    /// It also removes the given slots from `fully_available` and `partially_available`.
+    /// If the new ranges are identical to the current local slots, no change occurs and version
+    /// is not incremented.
+    ///
+    /// # Arguments
+    ///
+    /// * `ranges` - Slice of slot ranges. Must be sorted and normalized (no overlaps, no adjacent ranges).
+    pub fn set_local_slots(&mut self, ranges: &[SlotRange]) {
+        // Check if the ranges are already equal (no change needed)
+        if self.local == ranges {
+            return;
+        }
+
+        // Update local slots and remove from other sets
+        self.local.set_from_ranges(ranges);
+        self.fully_available.remove_ranges(ranges);
+        self.partially_available.remove_ranges(ranges);
+        self.increment_version();
+    }
+
+    /// Sets the partially available slot ranges.
+    ///
+    /// This function updates the `partially_available` set by adding the provided ranges.
+    /// It also removes the given slots from `local` and `fully_available`, and
+    /// increments the `version` counter (fully available or not available slots are now partially available).
+    ///
+    /// # Arguments
+    ///
+    /// * `ranges` - Slice of slot ranges. Must be sorted and normalized (no overlaps, no adjacent ranges).
+    pub fn set_partially_available_slots(&mut self, ranges: &[SlotRange]) {
```

Comment:
```suggestion
    /// Sets the partially available slot ranges.
    ///
    /// This function updates the `partially_available` set by adding the provided ranges.
    /// It also removes the given slots from `local` and `fully_available`, and
    /// increments the `version` counter (fully available or not available slots are now partially available).
    ///
    /// # Arguments
    ///
    /// * `ranges` - Slice of slot ranges. Must be sorted and normalized (no overlaps, no adjacent ranges).
    pub fn add_partially_available_slots(&mut self, ranges: &[SlotRange]) {
```
The other `set_*` function does a replace so using add instead avoids a confusion.